CppCoreGuidelines.md

```diff
@@ -5224,6 +5225,85 @@ Because that would in many cases -- especially simple cases -- be distracting cl
 
 (Simple) A destructor should be declared `noexcept` if it could throw.
 
+
+### <a name="Rc-dtor-final"></a>C.38: Define a class as `final` if it has a public non-virtual destructor
+
+##### Reason
+
+[A non-virtual destructor of base class should be protected](#Rc-dtor-virtual). So, the derived class is not allowed if a non-virtual destructor is public.
+
+##### Discussion
+
+See [this in the Discussion section](#Sd-dtor).
+
+##### Note
+
+Inheritance without virtuality can be done by extra base class. Declare the entire implementation in the base class and the default public non-virtual destructor in the child one. Please accept the base class but not the child one as an input parameter of your functions unless it is explicitly required for a child specific.
```

Comment:
> When this rule is added, shouldn't then also the good examples be adjusted, for example in https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#SS-concrete?

I do not see any example in SS-concrete with inheritance.
I would like to review an apply changes by separate commits.

Reply:
> It is not clear to me why the rule talks about a "child class".

My fault. Replaced to `derived`.

Reply:
> > Inheritance without virtuality can be done by extra base class. Declare the entire implementation in the base class and the default public non-virtual destructor in the child one.
> 
> This is something I would not immediately understand. Why would I want that? Is that something that I shall do? It looks like a suggestion. Also the placement here does not help. Next is a bad example with no relation to this Note. Maybe the note can be moved between the bad and the good example?

Agree. It is a suggestion on how to organize the class hierarchy within following to this rule. So, I movet this note before the next example and added some words about why and when it could be useful. I extended the example alse by cases with compilation failures.

I decided to move this rule into [C.hier: Class hierarchies (OOP)](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#SS-SS-hier) because it declares requrements for a class hierarhy, not for a destructor.

Reply:
> > Please accept the base class but not the child one as an input parameter of your functions unless it is explicitly required for a child specific.
> 
> The sentence seems to be incomplete. And to me it looks more like an extra rule, not like something that should be mentioned here.

It is common and often situation when final interface classes used as an imput parameters in library interfaces. It produces an issues for extension and plugins. I wanted to warn developers about such issue. But your are right, it should not be here. It could be a separate rule or part of some existing one. Perhaps the linkt to it will be ueful here.

So, I removed this sentence.

Reply:
[Workin](https://github.com/softogen/CppCoreGuidelines/commit/1a4699e8caa7504f84a03fd645cba828e9eca45d) on it separately. It is hard to make a decision in most of cases. The samples miss a lot of details to be a short, simple and focused on the specific topic only. Adding a final/virtual~/protected:~ produces a clutter. So, it isn't a simple and fast task to make all samples good enough.